# Imports

In [0]:
from pyspark.sql import functions as F
from pyspark.sql.window import Window
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import folium

In [0]:
df = spark.read \
    .format("csv") \
    .option("header", "true") \
    .option("inferSchema", "true") \
    .load("dbfs:/FileStore/tables/globalterrorismdb_0718dist_tar.bz2")

display(df.limit(10))
df.printSchema()


_c0 eventid iyear imonth iday approxdate extended resolution country country_txt region region_txt provstate city latitude longitude specificity vicinity location summary crit1 crit2 crit3 doubtterr alternative alternative_txt multiple success suicide attacktype1 attacktype1_txt attacktype2 attacktype2_txt attacktype3 attacktype3_txt targtype1 targtype1_txt targsubtype1 targsubtype1_txt corp1 target1 natlty1 natlty1_txt targtype2 targtype2_txt targsubtype2 targsubtype2_txt corp2 target2 natlty2 natlty2_txt targtype3 targtype3_txt targsubtype3 targsubtype3_txt corp3 target3 natlty3 natlty3_txt gname gsubname gname2 gsubname2 gname3 gsubname3 motive guncertain1 guncertain2 guncertain3 individual nperps nperpcap claimed claimmode claimmode_txt claim2 claimmode2 claimmode2_txt claim3 claimmode3 claimmode3_txt compclaim weaptype1 weaptype1_txt weapsubtype1 weapsubtype1_txt weaptype2 weaptype2_txt weapsubtype2 weapsubtype2_txt weaptype3 weaptype3_txt weapsubtype3 weapsubtype3_txt weaptype4 weaptype4_txt weapsubtype4 weapsubtype4_txt weapdetail nkill nkillus nkillter nwound nwoundus nwoundte property propextent propextent_txt propvalue propcomment ishostkid nhostkid nhostkidus nhours ndays divert kidhijcountry ransom ransomamt ransomamtus ransompaid ransompaidus ransomnote hostkidoutcome hostkidoutcome_txt nreleased addnotes scite1 scite2 scite3 dbsource INT_LOG INT_IDEO INT_MISC INT_ANY related 0 197000000001 1970 7 2 null 0 null 58 Dominican Republic 2 Central America & Caribbean null Santo Domingo 18.456792 -69.951164 1.0 0 null null 1 1 1 0.0 null null 0.0 1 0 1 Assassination null null null null 14 Private Citizens & Property 68.0 Named Civilian null Julio Guzman 58.0 Dominican Republic null null null null null null null null null null null null null null null null MANO-D null null null null null null 0.0 null null 0 null null null null null null null null null null null null 13 Unknown null null null null null null null null null null null null null null null 1.0 null null 0.0 null null 0 null null null null 0.0 null null null null null null 0.0 null null null null null null null null null null null null PGIS 0 0 0 0 null 1 197000000002 1970 0 0 null 0 null 130 Mexico 1 North America Federal Mexico city 19.371887 -99.086624 1.0 0 null null 1 1 1 0.0 null null 0.0 1 0 6 Hostage Taking (Kidnapping) null null null null 7 Government (Diplomatic) 45.0 Diplomatic Personnel (outside of embassy, consulate) Belgian Ambassador Daughter Nadine Chaval, daughter 21.0 Belgium null null null null null null null null null null null null null null null null 23rd of September Communist League null null null null null null 0.0 null null 0 7.0 null null null null null null null null null null null 13 Unknown null null null null null null null null null null null null null null null 0.0 null null 0.0 null null 0 null null null null 1.0 1.0 0.0 null null null Mexico 1.0 800000.0 null null null null null null null null null null null PGIS 0 1 1 1 null 2 197001000001 1970 1 0 null 0 null 160 Philippines 5 Southeast Asia Tarlac Unknown 15.478598 120.599741 4.0 0 null null 1 1 1 0.0 null null 0.0 1 0 1 Assassination null null null null 10 Journalists & Media 54.0 Radio Journalist/Staff/Facility Voice of America Employee 217.0 United States null null null null null null null null null null null null null null null null Unknown null null null null null null 0.0 null null 0 null null null null null null null null null null null null 13 Unknown null null null null null null null null null null null null null null null 1.0 null null 0.0 null null 0 null null null null 0.0 null null null null null null 0.0 null null null null null null null null null null null null PGIS -9 -9 1 1 null 3 197001000002 1970 1 0 null 0 null 78 Greece 8 Western Europe Attica Athens 37.99749 23.762728 1.0 0 null null 1 1 1 0.0 null null 0.0 1 0 3 Bombing/Explosion null null null null 7 Government (Diplomatic) 46.0 Embassy/Consulate null U.S. Embassy 217.0 United States null null nul

root
 |-- _c0: integer (nullable = true)
 |-- eventid: long (nullable = true)
 |-- iyear: integer (nullable = true)
 |-- imonth: integer (nullable = true)
 |-- iday: integer (nullable = true)
 |-- approxdate: string (nullable = true)
 |-- extended: integer (nullable = true)
 |-- resolution: date (nullable = true)
 |-- country: integer (nullable = true)
 |-- country_txt: string (nullable = true)
 |-- region: integer (nullable = true)
 |-- region_txt: string (nullable = true)
 |-- provstate: string (nullable = true)
 |-- city: string (nullable = true)
 |-- latitude: double (nullable = true)
 |-- longitude: double (nullable = true)
 |-- specificity: double (nullable = true)
 |-- vicinity: integer (nullable = true)
 |-- location: string (nullable = true)
 |-- summary: string (nullable = true)
 |-- crit1: string (nullable = true)
 |-- crit2: string (nullable = true)
 |-- crit3: string (nullable = true)
 |-- doubtterr: string (nullable = true)
 |-- alternative: string (nullable = true)
 |-- 

In [0]:
df = df.select(
    "eventid",
    F.col("iyear").alias("year"),
    F.col("imonth").alias("month"),
    F.col("iday").alias("day"),
    "country",
    "country_txt",
    "region",
    "region_txt",
    "provstate",
    "city",
    "latitude",
    "longitude",
    "success",
    "suicide",
    "attacktype1",
    "attacktype1_txt",
    "targtype1_txt",
    "targsubtype1_txt",
    "target1",
    "natlty1_txt",
    "gname",
    "gsubname",
    "nperps",
    "weaptype1_txt",
    "weapsubtype1_txt",
    "nkill",
    "nkillus"
)

In [0]:
display(df.limit(10))

eventid year month day country country_txt region region_txt provstate city latitude longitude success suicide attacktype1 attacktype1_txt targtype1_txt targsubtype1_txt target1 natlty1_txt gname gsubname nperps weaptype1_txt weapsubtype1_txt nkill nkillus 197000000001 1970 7 2 58 Dominican Republic 2 Central America & Caribbean null Santo Domingo 18.456792 -69.951164 1 0 1 Assassination Private Citizens & Property Named Civilian Julio Guzman Dominican Republic MANO-D null null Unknown null 1.0 null 197000000002 1970 0 0 130 Mexico 1 North America Federal Mexico city 19.371887 -99.086624 1 0 6 Hostage Taking (Kidnapping) Government (Diplomatic) Diplomatic Personnel (outside of embassy, consulate) Nadine Chaval, daughter Belgium 23rd of September Communist League null 7.0 Unknown null 0.0 null 197001000001 1970 1 0 160 Philippines 5 Southeast Asia Tarlac Unknown 15.478598 120.599741 1 0 1 Assassination Journalists & Media Radio Journalist/Staff/Facility Employee United States Unknown null null Unknown null 1.0 null 197001000002 1970 1 0 78 Greece 8 Western Europe Attica Athens 37.99749 23.762728 1 0 3 Bombing/Explosion Government (Diplomatic) Embassy/Consulate U.S. Embassy United States Unknown null null Explosives Unknown Explosive Type null null 197001000003 1970 1 0 101 Japan 4 East Asia Fukouka Fukouka 33.580412 130.396361 1 0 7 Facility/Infrastructure Attack Government (Diplomatic) Embassy/Consulate U.S. Consulate United States Unknown null null Incendiary null null null 197001010002 1970 1 1 217 United States 1 North America Illinois Cairo 37.005105 -89.176269 1 0 2 Armed Assault Police Police Building (headquarters, station, school) Cairo Police Headquarters United States Black Nationalists null -99.0 Firearms Unknown Gun Type 0.0 0.0 197001020001 1970 1 2 218 Uruguay 3 South America Montevideo Montevideo -34.891151 -56.187214 0 0 1 Assassination Police Police Security Forces/Officers Juan Maria de Lucah/Chief of Directorate of info. and intell. Uruguay Tupamaros (Uruguay) null 3.0 Firearms Automatic or Semi-Automatic Rifle 0.0 null 197001020002 1970 1 2 217 United States 1 North America California Oakland 37.791927 -122.225906 1 0 3 Bombing/Explosion Utilities Electricity Edes Substation United States Unknown null -99.0 Explosives Unknown Explosive Type 0.0 0.0 197001020003 1970 1 2 217 United States 1 North America Wisconsin Madison 43.076592 -89.412488 1 0 7 Facility/Infrastructure Attack Military Military Recruiting Station/Academy R.O.T.C. offices at University of Wisconsin, Madison United States New Year's Gang null 1.0 Incendiary Molotov Cocktail/Petrol Bomb 0.0 0.0 197001030001 1970 1 3 217 United States 1 North America Wisconsin Madison 43.07295 -89.386694 1 0 7 Facility/Infrastructure Attack Government (General) Government Building/Facility/Office Selective Service Headquarters in Madison Wisconsin United States New Year's Gang null 1.0 Incendiary Gasoline or Alcohol 0.0 0.0

In [0]:
df.printSchema()

root
 |-- eventid: long (nullable = true)
 |-- year: integer (nullable = true)
 |-- month: integer (nullable = true)
 |-- day: integer (nullable = true)
 |-- country: integer (nullable = true)
 |-- country_txt: string (nullable = true)
 |-- region: integer (nullable = true)
 |-- region_txt: string (nullable = true)
 |-- provstate: string (nullable = true)
 |-- city: string (nullable = true)
 |-- latitude: double (nullable = true)
 |-- longitude: double (nullable = true)
 |-- success: string (nullable = true)
 |-- suicide: string (nullable = true)
 |-- attacktype1: string (nullable = true)
 |-- attacktype1_txt: string (nullable = true)
 |-- targtype1_txt: string (nullable = true)
 |-- targsubtype1_txt: string (nullable = true)
 |-- target1: string (nullable = true)
 |-- natlty1_txt: string (nullable = true)
 |-- gname: string (nullable = true)
 |-- gsubname: string (nullable = true)
 |-- nperps: string (nullable = true)
 |-- weaptype1_txt: string (nullable = true)
 |-- weapsubtype1_txt:

success - Success of a terrorist strike

suicide - 1 = "Yes" The incident was a suicide attack. 0 = "No" There is no indication that the incident was a suicide

attacktype1 - The general method of attack

attacktype1_txt - The general method of attack and broad class of tactics used.

targtype1_txt - The general type of target/victim

targsubtype1_txt - The more specific target category

target1 - The specific person, building, installation that was targeted and/or victimized

natlty1_txt - The nationality of the target that was attacked

gname - The name of the group that carried out the attack

gsubname - Additional details about group that carried out the attack like fractions

nperps - The total number of terrorists participating in the incident

weaptype1_txt - General type of weapon used in the incident

weapsubtype1_txt - More specific value for most of the Weapon Types

nkill - The number of total confirmed fatalities for the incident

nkillus - The number of U.S. citizens who died as a result of the incident

#### Imagine you are a security or defense analyst. Analyze the data and draw conclusions on the distribution and nature of terrorist incidents recorded around the world. In your analysis, include maps that visualize the location of different incidents. Your analysis may also provide answers to the following questions:

# 1. How has the number of terrorist activities changed over the years? Are there certain regions where this trend is different from the global averages?

In [0]:
attacks_per_year = df.groupBy("year") \
    .agg(F.count("*").alias("total_attacks_global")) \
    .orderBy("year")
display(attacks_per_year.limit(20))

year total_attacks_global 1970 651 1971 471 1972 568 1973 473 1974 581 1975 740 1976 923 1977 1319 1978 1526 1979 2662 1980 2662 1981 2586 1982 2544 1983 2870 1984 3495 1985 2915 1986 2860 1987 3183 1988 3721 1989 4324

In [0]:
attacks_per_year_region = df.groupBy("year", "region_txt") \
    .agg(
        F.count("*").alias("total_attacks_region"),
    ) \
    .orderBy("year", "region_txt")

joined_df = attacks_per_year_region.join(attacks_per_year, on="year", how="left")
joined_df = joined_df.withColumn(
    "region_share",
    F.round((F.col("total_attacks_region") / F.col("total_attacks_global")) * 100, 1)
)


In [0]:
display(joined_df.limit(20))

year region_txt total_attacks_region total_attacks_global region_share 2002 Western Europe 120 1333 9.0 1981 Australasia & Oceania 3 2586 0.1 1984 South America 1492 3495 42.7 1975 Middle East & North Africa 44 740 5.9 1978 Eastern Europe 2 1526 0.1 1973 Western Europe 290 473 61.3 1980 Southeast Asia 87 2662 3.3 1980 Eastern Europe 1 2662 0.0 1982 Southeast Asia 43 2544 1.7 1991 Southeast Asia 206 4683 4.4 1976 North America 125 923 13.5 1979 Middle East & North Africa 455 2662 17.1 2000 Sub-Saharan Africa 191 1814 10.5 2003 East Asia 6 1278 0.5 1988 Sub-Saharan Africa 337 3721 9.1 1997 Central America & Caribbean 117 3197 3.7 1975 Western Europe 438 740 59.2 1977 Southeast Asia 8 1319 0.6 1978 South Asia 2 1526 0.1 1988 South Asia 801 3721 21.5

In [0]:
top_regions = df.groupBy("region_txt") \
    .agg(F.count("*").alias("total")) \
    .orderBy(F.desc("total")) \
    .limit(5)

top_regions_list = [row["region_txt"] for row in top_regions.collect()]

filtered_df = joined_df.filter(F.col("region_txt").isin(top_regions_list))

display(filtered_df.limit(20))

year region_txt total_attacks_region total_attacks_global region_share 2002 Western Europe 120 1333 9.0 1984 South America 1492 3495 42.7 1975 Middle East & North Africa 44 740 5.9 1973 Western Europe 290 473 61.3 1979 Middle East & North Africa 455 2662 17.1 2000 Sub-Saharan Africa 191 1814 10.5 1988 Sub-Saharan Africa 337 3721 9.1 1975 Western Europe 438 740 59.2 1978 South Asia 2 1526 0.1 1988 South Asia 801 3721 21.5 1974 South Asia 2 581 0.3 1991 Sub-Saharan Africa 271 4683 5.8 1994 South America 373 3456 10.8 1990 Western Europe 390 3887 10.0 1971 Middle East & North Africa 55 471 11.7 1982 Middle East & North Africa 290 2544 11.4 1983 Middle East & North Africa 334 2870 11.6 1985 Western Europe 466 2915 16.0 2002 Sub-Saharan Africa 121 1333 9.1 1992 South America 1204 5071 23.7

## 🌍 Global Trend Overview
From the global line chart:

- **1970s–1990s:** Gradual increase in terrorist incidents.
- **Late 1990s:** Noticeable decline.
- **2004–2014:** Sharp surge, peaking in 2014 with **~17,000** attacks.
- **Post-2014:** Decline, though attack levels remain historically high.

---

## 🌎 Regional Differences in Trends

### 🟠 Regions Following the Global Pattern

#### **Middle East & North Africa (MENA)**
- Major driver of global increase post-2004.
- Peaked at **~40–45%** of global share in 2014–2016.

#### **South Asia**
- Strong contributor to global rise, especially after 2005.
- Peaked at **~30–44%** between 2010 and 2014.

---

### 🔵 Regions Diverging from Global Pattern

#### **Western Europe**
- High share in **1970s–1980s** (e.g., 61% in 1973).
- Steady decline; **<5%** since 2000.

#### **North America**
- Dominated global share in **1970** (over 70%).
- Dropped significantly; now under **5%** post-1990.

#### **Central America & Caribbean**
- Peaked in the **1980s** (e.g., 40% in 1980).
- Minimal activity post-2000.

---

### 🟡 Regions with Moderate Trends

#### **Sub-Saharan Africa**
- Rising trend from mid-2000s.
- Peaked at **~15–18%** around 2014–2017.

#### **Southeast Asia**
- Consistent but moderate presence.
- Hovered around **5–10%**, with a bump post-2005.

---

## 📌 Summary Table

| Region                     | Matches Global Trend? | Key Observations |
|---------------------------|------------------------|------------------|
| Middle East & North Africa | ✅ Yes                 | Strong surge post-2004 |
| South Asia                | ✅ Yes                 | Peak around 2010–2014 |
| Western Europe            | ❌ No                  | High early on, major decline |
| North America             | ❌ No                  | Very high in 1970, now minimal |
| Central America & Caribbean | ❌ No               | Peaked in 1980s, near zero after 2000 |
| Sub-Saharan Africa        | ⚠️ Partially          | Moderate rise post-2005 |
| Southeast Asia            | ⚠️ Partially          | Stable, mild increase |
| South America             | ❌ No                  | High in 1980s, sharp fall afterward |


# 2. Is the number of incidents and the number of casualties correlated? Can you spot any irregularities or outliers?

In [0]:
yearl_stats = df \
  .groupBy("year") \
  .agg(
    F.count("eventid").alias("total_attacks"),
    F.sum("nkill").alias("total_killed")
  ) \
  .orderBy("year")

display(yearl_stats.limit(20))

year total_attacks total_killed 1970 651 174.0 1971 471 173.0 1972 568 564.0 1973 473 363.0 1974 581 530.0 1975 740 617.0 1976 923 674.0 1977 1319 456.0 1978 1526 1458.0 1979 2662 2099.0 1980 2662 4399.0 1981 2586 4851.0 1982 2544 5136.0 1983 2870 9443.0 1984 3495 10449.0 1985 2915 7091.0 1986 2860 4944.0 1987 3183 6482.0 1988 3721 7208.0 1989 4324 8152.0

In [0]:
correlation = yearl_stats.stat.corr("total_attacks", "total_killed")

print(f"Correlation between number of attacks and deaths: {correlation}")

Correlation between number of attacks and deaths: 0.965614695525471


In [0]:
display(yearl_stats \
    .withColumn(
        "lethality_ratio", F.col("total_killed") / F.col("total_attacks")
    ) \
    .orderBy(F.col("lethality_ratio").desc()) \
)

year total_attacks total_killed lethality_ratio 1998 934 4683.0 5.0139186295503215 2004 1166 5709.0 4.89622641509434 2001 1906 7706.0 4.0430220356768105 2007 3242 12749.0 3.932449105490438 2002 1333 4627.0 3.471117779444861 2006 2758 9333.0 3.3839738941261785 1997 3197 10815.0 3.3828589302471066 1983 2870 9443.0 3.290243902439024 2005 2017 6295.0 3.12097174020823 1984 3495 10449.0 2.989699570815451 2014 16903 44481.0 2.631544696207774 2015 14965 38815.0 2.5937186769127965 2016 13587 34763.0 2.5585486126444397 2003 1278 3257.0 2.548513302034429 1985 2915 7091.0 2.432590051457976 2017 10900 26408.0 2.4227522935779815 1999 1395 3367.0 2.413620071684588 2000 1814 4361.0 2.404079382579934 1996 3058 6966.0 2.277959450621321 1994 3456 7690.0 2.225115740740741 1987 3183 6482.0 2.0364436066603835 1982 2544 5136.0 2.018867924528302 1995 3081 6103.0 1.9808503732554366 2009 4721 9233.0 1.9557297182800255 1988 3721 7208.0 1.9371136791185166 1992 5071 9742.0 1.9211200946558864 1989 4324 8152.0 1.8852913968547642 2008 4805 9054.0 1.8842872008324663 1981 2586 4851.0 1.8758700696055683 2013 12036 22149.0 1.8402293120638085 1990 3887 7148.0 1.8389503473115514 1991 4683 8424.0 1.7988468930172965 2012 8522 15271.0 1.791950246421028 1986 2860 4944.0 1.7286713286713287 1980 2662 4399.0 1.6525169045830204 2011 5076 8146.0 1.6048069345941687 2010 4826 7575.0 1.5696228760878574 1972 568 564.0 0.9929577464788732 1978 1526 1458.0 0.9554390563564875 1974 581 530.0 0.9122203098106713 1975 740 617.0 0.8337837837837838 1979 2662 2099.0 0.7885048835462058 1973 473 363.0 0.7674418604651163 1976 923 674.0 0.7302275189599133 1971 471 173.0 0.3673036093418259 1977 1319 456.0 0.3457164518574678 1970 651 174.0 0.2672811059907834

In [0]:
high_lethality_years = [1998, 2001, 2004, 2007]

df_filtered = df.filter(F.col("year").isin(high_lethality_years))


attack_types = df_filtered.groupBy("year", "attacktype1_txt").agg(
    F.count("*").alias("attack_count"),
    F.sum("nkill").cast("float").alias("total_killed"),
    F.avg("nkill").cast("float").alias("avg_killed")
).orderBy("year", F.col("total_killed").desc())

display(attack_types)


year attacktype1_txt attack_count total_killed avg_killed 1998 Armed Assault 263 2018.0 8.73593 1998 Bombing/Explosion 502 1618.0 3.3568466 1998 Unknown 23 528.0 22.956522 1998 Hostage Taking (Kidnapping) 43 338.0 9.657143 1998 Unarmed Assault 15 101.0 6.733333 1998 Assassination 24 40.0 1.6666666 1998 Facility/Infrastructure Attack 48 39.0 0.8863636 1998 Hijacking 3 1.0 0.5 1998 1 2 null null 1998 3 1 null null 1998 2 1 null null 1998 0 3 null null 1998 7 6 null null 2001 Hijacking 14 3005.0 214.64285 2001 Armed Assault 580 2728.0 4.736111 2001 Bombing/Explosion 795 980.0 1.2612612 2001 Unknown 45 419.0 9.311111 2001 Hostage Taking (Kidnapping) 144 260.0 1.884058 2001 Assassination 126 144.0 1.152 2001 Facility/Infrastructure Attack 136 133.0 1.0075758 2001 Unarmed Assault 40 20.0 0.5405405 2001 Hostage Taking (Barricade Incident) 4 17.0 4.25 2001 6 1 null null 2001 1 1 null null 2001 3 1 null null 2001 0.0 6 null null 2001 0 5 null null 2001 7 4 null null 2001 null 3 null null 2001 2 1 null null 2004 Bombing/Explosion 656 3079.0 4.7223926 2004 Armed Assault 258 1764.0 6.890625 2004 Hostage Taking (Barricade Incident) 3 370.0 123.333336 2004 Unknown 26 195.0 7.5 2004 Assassination 103 169.0 1.6407768 2004 Hostage Taking (Kidnapping) 75 58.0 0.90625 2004 Hijacking 1 46.0 46.0 2004 Facility/Infrastructure Attack 32 28.0 0.875 2004 Unarmed Assault 4 0.0 0.0 2004 6 2 null null 2004 1 3 null null 2004 2 1 null null 2004 0.0 1 null null 2004 0 1 null null 2007 Bombing/Explosion 1860 8643.0 4.689636 2007 Armed Assault 848 3175.0 3.770784 2007 Hostage Taking (Kidnapping) 221 446.0 2.1339712 2007 Assassination 139 189.0 1.379562 2007 Unknown 30 161.0 5.366667 2007 Facility/Infrastructure Attack 113 117.0 1.1037736 2007 Hostage Taking (Barricade Incident) 6 10.0 2.5 2007 Unarmed Assault 6 7.0 1.1666666 2007 Hijacking 7 1.0 0.16666667 2007 0.0 3 null null 2007 null 1 null null 2007 2 1 null null 2007 1 3 null null 2007 0 4 null null

## ✅ Correlation Analysis

Using the yearly aggregated data:

- A **very strong positive correlation** of **0.9656** was found between the number of terrorist incidents (`total_attacks`) and the number of fatalities (`total_killed`).
- This confirms that, **in general, years with more terrorist attacks also saw more deaths**.

---

## ⚠️ Irregularities and Outliers

While the correlation is strong, an analysis of the **lethality ratio** (`total_killed / total_attacks`) highlights several years with disproportionately deadly attacks.


### 🔺 High Lethality Years
| Year | Lethality Ratio | Findings from `attacktype1_txt` |
|------|------------------|----------------------------------|
| **1998** | 5.01 | High lethality explained by a small number of extremely deadly **Bombings/Explosions**. |
| **2004** | 4.90 | Predominance of fatal **Bombings/Explosions**. |
| **2001** | 4.04 | Death toll dominated by **a few large-scale attacks**, most notably **9/11** (aviation-related suicide attacks). |
| **2007** | 3.93 | Many **Bombings/Explosions** and **Armed Assaults** with high casualty counts. |
| **2002–2006** | 3.3–3.5 | Consistently dominated by high-fatality types like **Bombings/Explosions** across multiple regions. |

These years demonstrate that **the lethality of a year depends not just on the number of attacks, but also on the type of attack carried out**.

---

## 🧠 Conclusion

- There is a strong statistical correlation between the number of terrorist attacks and total fatalities.
- However, **attack types (`attacktype1_txt`) play a critical role in explaining years with unusually high lethality**.



# 3. What are the most common methods of attacks? Does it differ in various regions or in time?

In [0]:
df_clean = df.filter(F.col("attacktype1_txt").isNotNull())


common_attacks = df_clean.groupBy("attacktype1_txt").agg(
    F.count("*").alias("total_attacks")
).orderBy(F.col("total_attacks").desc())


attacks_by_region = df_clean.groupBy("region_txt", "attacktype1_txt").agg(
    F.count("*").alias("total_attacks")
).orderBy("region_txt", F.col("total_attacks").desc())


region_window = Window.partitionBy("region_txt").orderBy(F.col("total_attacks").desc())
top_attacks_per_region = attacks_by_region.withColumn("rank", F.row_number().over(region_window)) \
    .filter(F.col("rank") <= 3)

attacks_by_year = df_clean.groupBy("year", "attacktype1_txt").agg(
    F.count("*").alias("total_attacks")
).orderBy("year", F.col("total_attacks").desc())


year_window = Window.partitionBy("year").orderBy(F.col("total_attacks").desc())
top_attacks_per_year = attacks_by_year.withColumn("rank", F.row_number().over(year_window)) \
    .filter(F.col("rank") <= 3)


print("🔹 Most common attack types overall:")
display(common_attacks)

print("🔹 Top 3 attack types per region:")
display(top_attacks_per_region.select("region_txt", "attacktype1_txt", "total_attacks"))

print("🔹 Top 3 attack types per year:")
display(top_attacks_per_year.select("year", "attacktype1_txt", "total_attacks"))

🔹 Most common attack types overall:


attacktype1_txt total_attacks Bombing/Explosion 88101 Armed Assault 42566 Assassination 19285 Hostage Taking (Kidnapping) 11115 Facility/Infrastructure Attack 10222 Unknown 7267 Unarmed Assault 1003 Hostage Taking (Barricade Incident) 989 Hijacking 657 0 130 1 85 3 56 7 52 2 40 0.0 37 6 16 1.0 15 8 5 9 4 5 2 and as such 1 he did manage to wound one of the robbers. The perpetrators escaped from the bank before they could steal any money." 1 "" which they translated into meaning 1 Mashkow 1 the Red Hand Defenders (RHD) 1 the choice 1 have been shifted to Peshawar via helicopters. Akhunzada Chattan 1 Other Crime Type 1 Anwar Abu Zaid 1

🔹 Top 3 attack types per region:


region_txt attacktype1_txt total_attacks Australasia & Oceania Bombing/Explosion 75 Australasia & Oceania Facility/Infrastructure Attack 70 Australasia & Oceania Armed Assault 51 Central America & Caribbean Armed Assault 4361 Central America & Caribbean Bombing/Explosion 3239 Central America & Caribbean Assassination 1253 Central Asia Bombing/Explosion 235 Central Asia Armed Assault 116 Central Asia Assassination 115 East Asia Bombing/Explosion 327 East Asia Facility/Infrastructure Attack 199 East Asia Armed Assault 117 Eastern Europe Bombing/Explosion 2761 Eastern Europe Armed Assault 1271 Eastern Europe Assassination 397 Middle East & North Africa Bombing/Explosion 30881 Middle East & North Africa Armed Assault 9254 Middle East & North Africa Assassination 4204 North America Bombing/Explosion 1507 North America Facility/Infrastructure Attack 821 North America Armed Assault 421 South America Bombing/Explosion 9024 South America Armed Assault 3874 South America Assassination 2744 South Asia Bombing/Explosion 21214 South Asia Armed Assault 11383 South Asia Assassination 4298 Southeast Asia Bombing/Explosion 4804 Southeast Asia Armed Assault 4003 Southeast Asia Assassination 1364 Sub-Saharan Africa Armed Assault 5995 Sub-Saharan Africa Bombing/Explosion 5552 Sub-Saharan Africa Hostage Taking (Kidnapping) 1867 Western Europe Bombing/Explosion 8482 Western Europe Assassination 2933 Western Europe Facility/Infrastructure Attack 2622

🔹 Top 3 attack types per year:


year attacktype1_txt total_attacks 1970 Bombing/Explosion 332 1970 Facility/Infrastructure Attack 173 1970 Armed Assault 61 1971 Bombing/Explosion 239 1971 Facility/Infrastructure Attack 88 1971 Assassination 70 1972 Assassination 265 1972 Bombing/Explosion 187 1972 Armed Assault 62 1973 Assassination 164 1973 Bombing/Explosion 149 1973 Armed Assault 54 1974 Bombing/Explosion 281 1974 Assassination 158 1974 Facility/Infrastructure Attack 42 1975 Bombing/Explosion 370 1975 Assassination 181 1975 Armed Assault 81 1976 Bombing/Explosion 419 1976 Assassination 204 1976 Armed Assault 124 1977 Bombing/Explosion 635 1977 Armed Assault 255 1977 Facility/Infrastructure Attack 181 1978 Bombing/Explosion 644 1978 Assassination 262 1978 Armed Assault 241 1979 Bombing/Explosion 1058 1979 Assassination 526 1979 Armed Assault 446 1980 Bombing/Explosion 997 1980 Assassination 618 1980 Armed Assault 574 1981 Bombing/Explosion 1082 1981 Armed Assault 697 1981 Assassination 405 1982 Bombing/Explosion 1125 1982 Armed Assault 665 1982 Assassination 361 1983 Bombing/Explosion 1245 1983 Armed Assault 852 1983 Assassination 360 1984 Bombing/Explosion 1775 1984 Armed Assault 823 1984 Assassination 443 1985 Bombing/Explosion 1479 1985 Armed Assault 659 1985 Assassination 311 1986 Bombing/Explosion 1506 1986 Armed Assault 591 1986 Assassination 371 1987 Bombing/Explosion 1477 1987 Armed Assault 798 1987 Assassination 495 1988 Bombing/Explosion 1649 1988 Armed Assault 922 1988 Assassination 821 1989 Bombing/Explosion 1796 1989 Armed Assault 1121 1989 Assassination 980 1990 Bombing/Explosion 1730 1990 Assassination 877 1990 Armed Assault 876 1991 Bombing/Explosion 1988 1991 Armed Assault 1271 1991 Assassination 730 1992 Bombing/Explosion 1738 1992 Armed Assault 1327 1992 Assassination 1111 1994 Bombing/Explosion 1153 1994 Armed Assault 816 1994 Assassination 771 1995 Bombing/Explosion 790 1995 Armed Assault 740 1995 Assassination 729 1996 Bombing/Explosion 1217 1996 Armed Assault 636 1996 Assassination 478 1997 Bombing/Explosion 1123 1997 Armed Assault 826 1997 Assassination 420 1998 Bombing/Explosion 502 1998 Armed Assault 263 1998 Facility/Infrastructure Attack 48 1999 Bombing/Explosion 646 1999 Armed Assault 366 1999 Hostage Taking (Kidnapping) 113 2000 Bombing/Explosion 892 2000 Armed Assault 491 2000 Hostage Taking (Kidnapping) 122 2001 Bombing/Explosion 795 2001 Armed Assault 580 2001 Hostage Taking (Kidnapping) 144 2002 Bombing/Explosion 712 2002 Armed Assault 366 2002 Hostage Taking (Kidnapping) 79 2003 Bombing/Explosion 676 2003 Armed Assault 313 2003 Assassination 87 2004 Bombing/Explosion 656 2004 Armed Assault 258 2004 Assassination 103 2005 Bombing/Explosion 1064 2005 Armed Assault 504 2005 Assassination 225 2006 Bombing/Explosion 1500 2006 Armed Assault 736 2006 Hostage Taking (Kidnapping) 196 2007 Bombing/Explosion 1860 2007 Armed Assault 848 2007 Hostage Taking (Kidnapping) 221 2008 Bombing/Explosion 2647 2008 Armed Assault 1083 2008 Hostage Taking (Kidnapping) 381 2009 Bombing/Explosion 2561 2009 Armed Assault 1104 2009 Facility/Infrastructure Attack 346 2010 Bombing/Explosion 2496 2010 Armed Assault 1108 2010 Hostage Taking (Kidnapping) 389 2011 Bombing/Explosion 2579 2011 Armed Assault 1394 2011 Hostage Taking (Kidnapping) 362 2012 Bombing/Explosion 5077 2012 Armed Assault 2082 2012 Assassination 458 2013 Bombing/Explosion 6679 2013 Armed Assault 2960 2013 Assassination 853 2014 Bombing/Explosion 8797 2014 Armed Assault 4069 2014 Hostage Taking (Kidnapping) 1396 2015 Bombing/Explosion 7660 2015 Armed Assault 3365 2015 Hostage Taking (Kidnapping) 1197 2016 Bombing/Explosion 7101 2016 Armed Assault 2716 2016 Hostage Taking (Kidnapping) 1129 2017 Bombing/Explosion 5017 2017 Armed Assault 2357 2017 Hostage Taking (Kidnapping) 877

## 🔍 What are the most common methods of terrorist attacks?
### Does it differ across regions or over time?

---

## ✅ Overall Most Common Attack Methods

The five most frequent attack methods across the entire dataset are:

| Attack Type                        | Total Attacks |
|-----------------------------------|----------------|
| **Bombing/Explosion**             | 88,101         |
| **Armed Assault**                 | 42,566         |
| **Assassination**                 | 19,285         |
| **Hostage Taking (Kidnapping)**   | 11,115         |
| **Facility/Infrastructure Attack**| 10,222         |

🔹 **Bombing/Explosion** is by far the dominant method globally, representing more than **40%** of all attacks.

---

## 🌍 Regional Differences

Attack methods vary significantly by region:

| Region                      | Top Attack Types (by frequency)                      |
|-----------------------------|-------------------------------------------------------|
| **Middle East & North Africa** | Bombing/Explosion, Armed Assault, Assassination       |
| **South Asia**                 | Bombing/Explosion, Armed Assault, Assassination       |
| **Sub-Saharan Africa**        | Armed Assault, Bombing/Explosion, Kidnapping          |
| **Central America & Caribbean** | Armed Assault, Bombing/Explosion, Assassination     |
| **Western Europe**            | Bombing/Explosion, Assassination, Facility Attacks    |
| **North America**             | Bombing/Explosion, Facility Attacks, Armed Assault    |
| **Eastern Europe**            | Bombing/Explosion, Armed Assault, Assassination       |
| **Southeast Asia**            | Bombing/Explosion, Armed Assault, Assassination       |
| **East Asia**                 | Bombing/Explosion, Facility Attacks, Armed Assault    |
| **Central Asia**              | Bombing/Explosion, Armed Assault, Assassination       |
| **Australasia & Oceania**     | Bombing/Explosion, Facility Attacks, Armed Assault    |

📌 While **Bombing/Explosion** leads in most regions, others like **Sub-Saharan Africa** emphasize **Armed Assaults** and **Kidnappings**, highlighting differences in tactics based on geography and conflict dynamics.

---

## 📆 Evolution Over Time

Temporal analysis of top attack types per year shows:

- **1970s–1980s**: Dominated by **Bombings**, **Assassinations**, and **Facility Attacks**.
- **1990s–early 2000s**: Continued dominance of **Bombings**, but with growing use of **Armed Assaults**.
- **2010–2017**: Surge in **Hostage Takings (Kidnapping)** alongside **Bombings** and **Armed Assaults**.

📈 The use of **Bombing/Explosion** remains consistently the most used method throughout the decades.

---

## 🧠 Conclusion

- **Bombing/Explosion** is the most common method of terrorist attack globally, across time and regions.
- **Significant regional variation** exists: e.g., **Armed Assaults dominate in Sub-Saharan Africa**, while **Facility Attacks** are more common in North America and East Asia.
- Over time, **Hostage Takings and Armed Assaults** have become more prominent, especially post-2000.

# 4. Plot the locations of attacks on a map to visualize their regional spread

In [0]:

df_geo = df.filter(
    (F.col("latitude").isNotNull()) & 
    (F.col("longitude").isNotNull())
).select("eventid", "year", "region_txt", "country_txt", "city", "latitude", "longitude")


df_geo_sampled = df_geo.limit(10000)

pdf_geo = df_geo_sampled.toPandas()

In [0]:
fig = px.scatter_geo(
    pdf_geo,
    lat="latitude",
    lon="longitude",
    color="region_txt",
    hover_name="country_txt",
    hover_data=["city", "year"],
    title="Global Terrorist Attack Locations",
    opacity=0.6,
    height=600
)

fig.update_layout(geo=dict(showland=True, showcountries=True))
fig.show()